### PR TITLE
handle RSA key too small and consume errors on CSR signature failure

### PR DESCRIPTION
**Pull request by reaperhulk**
This PR is not ready for merge as I'm not sure how to trigger the generic failure case (e.g. signature failure that isn't RSA key too small). Suggestions?

### CODE REVIEW

File: src/cryptography/hazmat/backends/openssl/backend.py [hunk 1]
```diff
@@ -1055,7 +1055,19 @@ def create_x509_csr(self, builder, private_key, algorithm):
...
-        assert res > 0
+        if res == 0:
+            errors = self._consume_errors()
+            if errors[0][1:] in (
+                (
+                    self._lib.ERR_LIB_RSA,
+                    self._lib.RSA_F_RSA_SIGN,
+                    self._lib.RSA_R_DIGEST_TOO_BIG_FOR_RSA_KEY
+                ),
+            ):
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/backend.py]: Just make this an assert?

Comment (reaperhulk) [inline src/cryptography/hazmat/backends/openssl/backend.py]: Derp, good idea.

File: src/cryptography/hazmat/backends/openssl/backend.py [hunk 2]
```diff
@@ -1055,7 +1055,16 @@ def create_x509_csr(self, builder, private_key, algorithm):
         res = self._lib.X509_REQ_sign(
             x509_req, private_key._evp_pkey, evp_md
         )
-        assert res > 0
+        if res == 0:
+            errors = self._consume_errors()
+            assert errors[0][1:] in (
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/backend.py]: Is there a reason to use `in` instead of `==`?

Comment (reaperhulk) [inline src/cryptography/hazmat/backends/openssl/backend.py]: Nope, left it in there from when I was looking for more than one error condition.

File: src/cryptography/hazmat/backends/openssl/backend.py [hunk 3]
```diff
@@ -1055,7 +1055,16 @@ def create_x509_csr(self, builder, private_key, algorithm):
...
-        assert res > 0
+        if res == 0:
+            errors = self._consume_errors()
+            assert errors[0][1:] == (
+                (
+                    self._lib.ERR_LIB_RSA,
+                    self._lib.RSA_F_RSA_SIGN,
+                    self._lib.RSA_R_DIGEST_TOO_BIG_FOR_RSA_KEY
+                )
+            )
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/backend.py]: You can drop the outer parens I guess.

File: tests/test_x509.py [hunk 4]
```diff
@@ -1206,6 +1206,19 @@ def test_extended_key_usage(self, backend):
...
+        private_key = rsa.generate_private_key(65537, 512, backend)
+        builder = x509.CertificateSigningRequestBuilder()
+        builder = builder.subject_name(
+            x509.Name([x509.NameAttribute(x509.OID_COUNTRY_NAME, u'US')])
+        )
+
+        with pytest.raises(ValueError) as exc:
+            builder.sign(private_key, hashes.SHA512(), backend)
+
+        assert exc.value.message == "Digest too big for RSA key"
```
Comment (alex) [inline tests/test_x509.py]: No `.message` on py3k, use `str(exc.value)` I guess

**Pull request merged**